### ISSUE TITLE
Improve model evaluate and __call__ API

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#2443]
We should improve our API pattern for model evaluation.

Sometimes we have `__call__` (especially for spectral models) and call that from Gammapy, sometimes we don't have it or don't use it (I think for spatial and temporal models).

Sometimes `evaluate` takes `self` (e.g. in [AbsorbedSpectralModel](https://docs.gammapy.org/0.14/_modules/gammapy/modeling/models/spectral.html#AbsorbedSpectralModel), when the model has a "data" attribute other than the parameters), sometimes we use `@staticmethod`. This kind of makes sense, but if we keep this we should add a second example how to do this, i.e. move [this example](https://docs.gammapy.org/0.14/notebooks/spectrum_simulation.html#Adding-a-user-defined-model) to `tutorials/models.ipynb`, and add a second one with a model that has data, or at least point to `AbsorbedSpectralModel` as an example.

We now have several model-related issues open: #2353, #2437, but I wanted to note this specific issue concerning "models with data" since it came up today.

Generally I think the way forward is that @adonath will close the old modeling PIG (#1971), and then @adonath and I will write a small and short new PIG early next week that proposes concrete solutions for the remaining open modeling-related issues, so that we have a plan how to get to v1.0 quickly.

But @adonath or @registerrier or anyone -- if you have suggestions on specific points now, feel  free to put them as comments in Github issues any time.

**Comment by cdeil** [comment]
Another note about https://docs.gammapy.org/0.14/notebooks/spectrum_simulation.html#Adding-a-user-defined-model

We should remove `SpectralModel.__subclasses__()` and instead show the model registries (see #2387). I looked into `SpectralModel.__subclasses__()` recently and it's recommended to regard it as an implementation detail of CPython, it shouldn't be used or relied upon.

Also, while on the topic, most likely the way to improve the model subclass and to make parameters descriptors is to use `__init_subclass__` and `__set_name__`, which are mentioned here: https://www.python.org/dev/peps/pep-0487/
This is a Python 3.6+ feature that is a perfect fit for us to make a little modeling framework and to do nicer parameter handling, without requiring the introduction of a metaclass or class decorator for that purpose. Implementing something very similar to how `astropy.modeling` works is now possible I think with just a little bit of code in `Model.__init_subclass__`. It will still be kind of the most advanced code in Gammapy, but I think it'll be acceptable to have ~ 20 lines of code to do some metaprogramming in `Model.__init_subclass__`, it requires a good understanding how Python classes work, but it's not magic.

**Comment by cdeil** [comment]
Another thing I noticed: there's a `_convert_energy` with hard-coded parameter names `reference` and `emin` which was added in https://github.com/gammapy/gammapy/commit/8363d93f788aa3ad58c4a6fb801eb6fb3be8e673#diff-742ab580dbb09ffcf34d43332a8a15aaR62-R67 to make evaluate error, and integral evaluate make for 3FGL catalogs.

If I delete it, only some tests in `TestFermi3FGLObject` fail, there's no unit test for `SpectralModel`. It's probably pretty error-prone, some concerns: what if the parameter names `reference` or `emin` change? When if `emax` and `emin` have different units? what if someone overwrites one of the methods in the base class that call `_convert_energy` and forgets to add it? What about 3D models, e.g. if we add an `EnergyDependentGaussian3D` - ideally energy evaluate unit handling would work the same there without duplicating quantity handling?

@adonath - Any ideas how to make this better?

Maybe the energy parameter handling for all evaluation methods (`evaluate`, `evaluate_error`, `integral`) can be handled in a generic way that doesn't rely on model parameter names?
This seems to be the case for `__call__` already:

https://github.com/gammapy/gammapy/blob/699c69e7ee52ae499b8ccc929f3bbf8e19b80afb/gammapy/modeling/models/spectral.py#L19-L27

Maybe ``__call__`` and ``_convert_energy`` can be unified?

**Comment by adonath** [comment]
@cdeil Yes, let's continue the discussion from https://github.com/gammapy/gammapy/pull/2255#issuecomment-553298060 here. 

The approach we tried to follow for spectral models is to introduce a translation layer on the `SpectralModel` base class, but try to keep the `.evaluate()` methods compatible with numpy arrays. This means avoid any `.to()` or `.to_value()` calls in `.evaluate()`, because writing pure numpy expressions, will just make it work for `Quantity` as well. But there is one exception: expressions like `(energy / energy_ref) ** index`, are numerically instable, when `energy` and `energy_ref` have different units. There are now two possible solutions: either call explicitly `(energy / energy_ref).to("")`, which fails for numpy arrays, or convert the energy units in a translation layer, as we do now in `__call__` and `_convert_energy()`. I think once PIG 14 is implemented the need for `_convert_energy()` goes away.

In general I think this is still a good approach. So we have `.evaluate()` methods on our models, that have to be compatible with numpy and must no implement any quantity specific code. Then we have a translation layer in the base class, where we introduce the support for `Quantity`, `SkyCoord` or `Time`. For the model evaluation during fitting we currently also go trough `__call__`, which is fine and can remain like this.

It's clear that for covariance and gradients we can not support `Quantity` objects, because they don't have homogeneous units. Still we can try to follow a similar approach. I.e. have e.g. an `evaluate_error()` method on the base class, which support units and have a lower-level `.evaluate_error_np()` method, which supports numpy.

Probably we should discuss this at the coding sprint and write up a minimal proposal.

**Comment by cdeil** [comment]
@adonath - Yes, but in reality it's more complex. It's probably a good idea to have a two-layered `__call__` and `evaluate`. But what about the other very important use cases of `evaluate_error` and `integrate` (on energy grid or map geom)? Do we also have a two-layer approach there? Is the quantity handling layer shared or different with `__call__`? Another issue is that I think we have random combinations of going though `evaluate` or `__call__` from the ouside, and even if you say `evaluate` should work just with arrays, I think mostly we test it for `Quantity` inputs only, there's no systematic test showing if it works for arrays or quantities or both for all models.

I don't spell it out in this PIG and point to the bad code, but that's the problem with the current spectral evaluation code, there's no clear concept yet or uniform implementation for how `evaluate_error` and `integral` work (and for spatial and time models it's just OK because we don't support error propagation yet, otherwise it would get similarly complex).

> Probably we should discuss this at the coding sprint and write up a minimal proposal.

Yes. But to be effective someone has to prepare the discussion and try to come already with a proposal. Can you please put a 5+5 min presentation + discussion on "model evaluation" for Monday?
If you want to do it, great! Otherwise you can put me.
I think it should be separate from the error propagation and this PIG, I decided to limit the scope of this PIG to not become the new model PIG where model evaluation and integration is designed.

**Comment by cdeil** [comment]
PS: I think the most effective way at this point is to send a PR with a modeling docs section with a few sentences or bullet points describing how evaluation and integration should work. Then we discuss this, and once we agree, we clean up our code and make it match. That's more effective than slides, because then we already have the docs in place.

**Comment by registerrier** [comment]
Closing.
